massa-execution/src/worker.rs

```diff
@@ -48,52 +55,78 @@ pub struct ExecutionWorker {
     ordered_active_blocks: Vec<(BlockId, Block)>,
     /// pending CSS final blocks
     ordered_pending_css_final_blocks: Vec<(BlockId, Block)>,
-    /// VM
+    /// VM thread
     vm_thread: JoinHandle<()>,
     /// VM execution requests queue
     execution_queue: ExecutionQueue,
 }
 
 impl ExecutionWorker {
     pub fn new(
-        cfg: ExecutionConfig,
+        cfg: ExecutionSettings,
         thread_count: u8,
         event_sender: mpsc::UnboundedSender<ExecutionEvent>,
         controller_command_rx: mpsc::Receiver<ExecutionCommand>,
         controller_manager_rx: mpsc::Receiver<ExecutionManagementCommand>,
+        bootstrap_state: Option<BootstrapExecutionState>,
     ) -> Result<ExecutionWorker, ExecutionError> {
         let execution_queue = ExecutionQueue::default();
         let execution_queue_clone = execution_queue.clone();
         let cfg_clone = cfg.clone();
-        // Start vm thread
+
+        // Check bootstrap
+        let bootstrap_final_slot;
+        let bootstrap_ledger;
+        if let Some(bootstrap_state) = bootstrap_state {
+            // init from bootstrap
+            bootstrap_final_slot = bootstrap_state.final_slot;
+            bootstrap_ledger = Some((bootstrap_state.final_ledger, bootstrap_final_slot));
+        } else {
+            // init without bootstrap
+            bootstrap_final_slot = Slot::new(0, thread_count.saturating_sub(1));
+            bootstrap_ledger = None;
+        };
+
+        // Init VM
+        let vm = Arc::new(Mutex::new(VM::new(cfg, thread_count, bootstrap_ledger)?));
+        let vm_clone = vm.clone();
+
+        // Start VM thread
         let vm_thread = thread::spawn(move || {
-            let mut vm = VM::new(cfg);
             let (lock, condvar) = &*execution_queue_clone;
             let mut requests = lock.lock().unwrap();
             requests.push_back(ExecutionRequest::Starting);
             // Run until shutdown.
             loop {
                 match &requests.pop_front() {
-                    Some(ExecutionRequest::RunFinalStep(step)) => vm.run_final_step(step),
-                    Some(ExecutionRequest::RunActiveStep(step)) => vm.run_active_step(step),
-                    Some(ExecutionRequest::ResetToFinalState) => vm.reset_to_final(),
+                    Some(ExecutionRequest::RunFinalStep(step)) => {
+                        vm_clone.lock().unwrap().run_final_step(step)
+                    }
+                    Some(ExecutionRequest::RunActiveStep(step)) => {
+                        vm_clone.lock().unwrap().run_active_step(step)
+                    }
+                    Some(ExecutionRequest::ResetToFinalState) => {
+                        vm_clone.lock().unwrap().reset_to_final()
+                    }
```

Comment:
big code for big hands? It isn't more readable at all

Reply:
any idea for a more readable alternative ?

Reply:
without braces :)

Reply:
cargo fmt does not share your aesthetic standards (it puts them back) :)

Reply:
ok! good for me :beers: